### ISSUE TITLE
Refactor keyboard help content key icons to have consistent look/feel

### ISSUE THREAD

**Issue by zepumph** [phetsims/scenery-phet#343]
In https://github.com/phetsims/friction/issues/45 I addressed a problem that I had in hard coding a wasd row to look just like the arrow key rows that are being added to the keyboard help dialogs for keyboard navigation. I would love to create a system so that the default options of what a "key" looks like can be passed through everywhere it's needed, but without creating circular dependencies.

https://github.com/phetsims/friction/issues/45#issuecomment-340121047 expresses how I hard coded things, and it doesn't look very great, especially from a maintainability perspective. @jessegreenberg what do you think?

**Comment by jessegreenberg** [comment]
@zepumph we could have something like LetterKeyNode type that manages the default options and layout for these icons.

One of the problems in phetsims/friction#45 was that the letter icons and arrow icons needed to align, and a scaling values was added by inspection.

![capture](https://user-images.githubusercontent.com/6396244/32246538-e40f566a-be44-11e7-8f41-eac66c78f238.PNG)

This will appear many places, maybe we should have a helper function that creates this and makes sure the arrow keys and WASD keys align?

**Comment by zepumph** [comment]
>This will appear many places, maybe we should have a helper function that creates this and makes sure the arrow keys and WASD keys align?

I would go one step further. I think that we should have a parent key node type that encompasses all "normal" sized keys. Currently `TextKeyNode` handles "normal" and "oversized" keys based on the text input, but I think something like `EscapeKeyNode` should extend from the same place as an arrow key or letter key such that the formatting is handled in the same place. **This way we have consistent sizing of square keys**. 

On top of that I think that there is styling of the key which applies to all keys, like shadowing. Currently the implementation of that is in `KeyNode`, which seems right, but `TextKeyNode` overrides that, so our arrows and our text keys are default different. 


Maybe the heirarchy looks like:

* KeyNode --- has all of the formatting for shadow, as well as height (since that should be consistent for all keys)
   * SquareKeyNode
         * SquareTextKeyNode
         * ArrowKeyNode
   * WideTextKeyNode

This is annoying though since TextKeyNode has a fair bit of logic. I'm still not sure what is best here.

**Comment by jessegreenberg** [comment]
> I think that we should have a parent key node type that encompasses all "normal" sized keys

I agree, that is kind of what I was thinking with `LetterKeyNode`, maybe that name is not general enough.

> This way we have consistent sizing of square keys.

In order to force this, would the type have to scale down the icon that is passed into it?

**Comment by zepumph** [comment]
Discussed today in meeting, I am going to take a stab at centralizing style in KeyNode, and then having KeyNode take in a node, and then scale it to the size to make it consistent with all keys.

**Comment by zepumph** [comment]
I updated the hierarchy so that KeyNode is responsible for a consistent height and the shadow. I also added all of the HelpContent features to panels to a demo in scenery phet on the components screen. Here is the link I use to get at it easily http://localhost/scenery-phet/scenery-phet_en.html?brand=phet&ea&screens=3&component=HelpContent
And here is a live screenshot!

![image](https://user-images.githubusercontent.com/6856943/32679830-5818daf0-c614-11e7-8153-c65c3b99e4c9.png)


With no options provided to any of these, they all have the same shadow and height. There is still work to be done though. 

- [ ] We need to make sure that the height isn't an option, but rather a constant and assert that no one ever tries to get at the height of the KeyNode in the constructor unless they want trouble. 
- [ ] There are some TODOs that I added around to make sure I don't forget some clean up
- [ ] All of the current dialogs need have size/scale/shadow styling taken away from them so that everything can be adjusted in a single place.

I'll keep at this for a few more hours, I just wanted to give an update. @jessegreenberg thanks for your help on DAG layout stuff.

**Comment by zepumph** [comment]
I think this is ready for review. @jessegreenberg could you take a look at some commits and some of the HelpContent implementations (both in scenery-phet demos and in sims) and see how we did. Please make recommendations for next steps.

**Comment by jessegreenberg** [comment]
@zepumph this is looking great, far fewer magic numbers and duplicated options.

In https://github.com/phetsims/friction/issues/45 @terracoda mentioned that the W key is a bit wider than the others. I tried using a combination of `keyWidthMargin` and `keyHeighMargin` to make letter keys consistent, but noticed that `keyWidthMargin` and `keyHeightMargin` behave differently.

`keyWidthMargin` gives keys additional padding by making the keys wider, while `keyHightMargin` gives more padding by scaling down the icon. I think these options should behave the same and both should scale down the icon so that it is easy to make sets of keys with the same dimensions. Thoughts?

**Comment by jessegreenberg** [comment]
On second thought,  maybe we need both options available to support default square keys. Maybe we need something like:
```
// x and y margins for the icon in the key, will translate the icon within the key (handled by AlignGroup)
xMargin: number
yMargin: number

// x and y padding, making the key larger by adding additional space around the icon
xPadding: number
yPadding: number

// scales up/down the icon to width and height so it is easy to make consistently sized keys
iconWidth: number|null
iconHeight: number|null

```

**Comment by zepumph** [comment]
I like your suggestions. I have been thinking about this implementation, and struggling. I feel like there are two cases and you already basically mentioned them:

1. You want to specify padding to make sure that the KeyNode looks to surround the icon, but you don't really care what the key size is.
2. You really want all the keys to be the same size, and the icons should scale. 

I was adding a level of complexity also because I was saying that you could do both for width, but only (1) for height. 

I'm going to go for an implementation that is more flexible using your options. I was thinking that I would do one change though. Would it be better to specify the width/height of the KeyNode rather than the icon?

**Comment by jessegreenberg** [comment]
> Would it be better to specify the width/height of the KeyNode rather than the icon?

That does seem more direct since it is the intended usage. Otherwise client will have to calculate width from icon width, padding options, and margins.

**Comment by zepumph** [comment]
I added support for a "forced" square key, that will scale the height down as well as scale the width down. I then increased the size of the keys just a hair, and reduced the padding. How does this look @jessegreenberg @terracoda 
![image](https://user-images.githubusercontent.com/6856943/32812796-559e0038-c953-11e7-99cb-74d06c2bc634.png)

**Comment by zepumph** [comment]
Also in master.

**Comment by zepumph** [comment]
Here is a better look at the design of just the Keys:
http://localhost/scenery-phet/scenery-phet_en.html?brand=phet&ea&screens=3&component=HelpContent

**Comment by zepumph** [comment]
![image](https://user-images.githubusercontent.com/6856943/32813499-03ab58b2-c957-11e7-9478-c65ef5b1c6f9.png)

![image](https://user-images.githubusercontent.com/6856943/32813509-0fb212fe-c957-11e7-8da0-b34e1eb42912.png)

**Comment by jessegreenberg** [comment]
@zepumph the square keys look great, nice job! Im also happy to review code changes, please reassign to me when ready.

**Comment by zepumph** [comment]
Ready for a code review!

**Comment by jessegreenberg** [comment]
@zepumph could `forceSquareKey` be a default option for ArrowKeyNode? Made the change in the above commit, but we can revert if that is problematic.

**Comment by zepumph** [comment]
Sounds nice and makes sense. I totally trust you.

**Comment by jessegreenberg** [comment]
I got a little confused by the xPadding and yPadding options, it seems like they behave like this:
```
xPadding:
if ( !forceSquareKey ) { make key wider to add padding }
if ( forceSquareKey ) { scale icon down to add padding }

yPadding:
always scale icon down to make padding
```
I think that making the key wider with `xPadding` is redundant anyway since that can be done with `minKeyWidth`.  So what if the options were like this?

```js
// x and y padding, always scaling down the icon to create padding
xPadding: 3,
yPadding: 3,

// Key will be at least this wide, making it possible to surround the icon with extra space if necessary.
// The minimum width of the KeyNode allowed, if the icon is wider, than it will expand gracefully
minKeyWidth: 17, // default equal to the height, a square key as the minimum.

// the desired height of the KeyNode; icon will be scaled down if too big
keyHeight: 17,

// Force the width of the KeyNode to be the same width as height, based on the height.
// Will scale down the icon if too wide.
forceSquareKey: false
```

**Comment by jessegreenberg** [comment]
Wait, sorry, that makes it impossible to create x padding around an icon during construction,  you would always have to pass in a minKeyWidth at least as large as your icon width if you want the key to look good. Scratch https://github.com/phetsims/scenery-phet/issues/343#issuecomment-345069173.

**Comment by jessegreenberg** [comment]
I udpated the documentation for the options based on their  behavior in https://github.com/phetsims/scenery-phet/issues/343#issuecomment-345069173.

**Comment by jessegreenberg** [comment]
Changes look great @zepumph. In the end, I added documentation for xPadding and yPadding options, changed some default values to match mockups (https://github.com/phetsims/molecules-and-light/issues/160), and created a LetterKeyNode type with default options for single letter keys, since it turns out they need slightly different margins. @zepumph could you please let me know if you have thoughts about those changes?

**Comment by zepumph** [comment]
This is looking very nice! Thanks for the skilled pixel polishing and organization. `LetterKeyNode` was a really nice addition. I think we are ready to close. @jessegreenberg can you think of anything else to be done here?

**Comment by jessegreenberg** [comment]
I don't think so @zepumph thanks for working on this issue, our keys are much better off. Closing.
